ddtrace/context.py

```diff
@@ -138,72 +135,26 @@ def add_span(self, span):
         Add a span to the context trace list, keeping it as the last active span.
         """
         with self._lock:
+            # Assume the first span added to the context is the local root
+            if self._local_root_span is None:
+                self._local_root_span = span
             self._set_current_span(span)
-            self._trace.append(span)
             span._context = self
 
     def close_span(self, span):
-        # type: (Span) -> Tuple[Optional[List[Span]], Optional[bool]]
+        # type: (Span) -> None
         """
         Mark a span as a finished, increasing the internal counter to prevent
         cycles inside _trace list.
         """
         with self._lock:
-            self._finished_spans += 1
-
-            # Safe-guard: prevent the last current span from being set to the parent
-            # of any span but the top-level span.
-            # The situation this avoids is when a parent closes before a child
-            # and the child is the last to close in the trace. When this happens
-            # the current_span would otherwise be set to the child's parent which
-            # has already closed. The context will be reset but the current_span
-            # will still point to that child's parent which would cause subsequent
-            # spans to be parented incorrectly.
-            if self._finished_spans != len(self._trace) or span == self._trace[0]:
-                self._set_current_span(span._parent)
-
-            if self._finished_spans == len(self._trace):
-                # get the trace
-                trace = self._trace
-                sampled = self._is_sampled()
-                sampling_priority = self._sampling_priority
-                # attach the sampling priority to the context root span
-                if sampled and sampling_priority is not None and trace:
-                    trace[0].set_metric(SAMPLING_PRIORITY_KEY, sampling_priority)
-                origin = self.dd_origin
-                # attach the origin to the root span tag
-                if sampled and origin is not None and trace:
-                    trace[0].meta[ORIGIN_KEY] = str(origin)
-
-                # clean the current state
-                self._trace = []
-                self._finished_spans = 0
+            if span == self._local_root_span:
```

Comment:
Add either a comment here or some refactoring to make clear why these attributes only apply to the local root span.

Reply:
Hm there's no change here from the previous implementation but we can add it since it was lacking before